### ISSUE TITLE
Ignoring specific keys in all dictionaries during validation and conversion

### ISSUE THREAD

**Issue by prescod** [samuelcolvin/pydantic#1273]
# Question

# Feature Request

Thank you for Pydantic. It seems great so far!

I have extended my YAML parser to keep track of the filename and line number where each dictionary comes from. Using this information in Pydantic has proven quite a challenge, which isn't surprising, because it presumably was not a use-case you initially considered.

There are two basic ways I can think of to represent my line number information.

1. In dictionaries:

`{"key1": "value1", "key2", "value2", "__linenumber__": 25}`

The question is how to data-type this? Rather than the clean and straightforward Dict[str, str] I now have to include a Union.

This makes the models very messy. Forcing the people who use my Yaml+Linenumbers_+Pydantic library to deal with this messiness is non-optimal.

2. In a dictionary subclass. In this case, the dictionary keys and values are heterogenous but the dictionary object has an additional attribute called `dict.__linenumber__` on it. This fixes the case of objects data-typed as Dictionaries, but it breaks objects that map to models. The signature of the model `__init__` is `__init__(**kwargs)`. So the model object never sees the dictionary object and can't grab the line number information from it. For my use-case it would be more convenient if the object itself was the argument to `__init__`.

I have a monkey-patch to ignore the magic keys, but I thought I would check with you about whether there is a better solution or if there might be in the future.

**Comment by samuelcolvin** [comment]
I agree there's no obvious workaround.

What's the use case here? Is it just for error reporting, or do you want to do something else with the line numbers?

If it's for something other than error reporting, could you please provide some context on the usage?

If it is just for error reporting, it's quite a common case so I could imagine making some changes to pydantic to make this easier.

**Comment by prescod** [comment]
Yes exactly: for error reporting. When I had a hand-rolled validator I would print filenames and line numbers with each error. It was a fair amount of work to convince PyYaml to keep the error messages and I was so close to getting it working with Pydantic. I'd even figured out how to prepend it to the _loc tuple.

There is no real standard for how json and yaml parsers would encode line number information so I can't offer you a silver bullet rule to follow of what to ignore. I personally had tried two conventions.

1. A dict-key named `__lineinfo__`

2. An instance attribute on a custom mapping type called `mapping.__lineinfo__`

In my own project I extended PyYAML to include those.

For my personal use-case, one technique that would work is if I could define a key name to be excluded from validation globally, but preserved to be captured in `__init__`.

**Comment by samuelcolvin** [comment]
Currently the best solution would probably be something like (untried):

```py
class MyModel(BaseModel):
    ...

    def __init__(self, **data):
        lineinfo = data.pop('lineinfo')
        super().__init__(**data)
        object.__setattr__(self, 'lineinfo', lineinfo)
```
(see #655 for a longer discussion of setting attributes on models)

---

Regarding a more completely approach to communicating more information to errors from the input data - I have no clear idea how this would work, but I would imagine some way (possibly another argument to `parse_obj()`) that let's you extract line info from part of an object to include in an error.

But it's going to be hyper general, not specific to a particular library or even easy to share between libraries, since as you say, there's no standard way of including line info with data.

perhaps related to #982

**Comment by prescod** [comment]
I've gone round and round on solutions like that.

The problem is that my yaml parser adds the lineinfo to *every* dictionary, including those of type Dict[str, str], Dict[str, int] and so forth. The parser doesn't know which dictionaries represent maps and which represent model objects.

I think one way it could have worked for me is if I could have overridden __VALIDATORS[dict] and told it how to exclude my `extra` attributes. I think that worked when I hacked it and I can verify it if you think you might be interested in a solution of that nature. That's why I had included "ignoring specific keys" in my original bug report. Of course there are many ways to skin a cat and excluding keys is just one way. So let's call that proposal 1: Excluding Keys

Proposal 2: If you passed the ORIGINAL object to my `__init__` as __orig then I could smuggle in the information I need as an attribute instead of a key. Which wouldn't mess up my dictionary processing code.

Proposal 3: How about if there were a general concept of a pre/post-validation or pre/post-instantiation hooks where I can add or remove things on the dictionary before it is validated or instantiated. Then I can implement whatever weird logic makes sense for my application.

This one is more involved but maybe its the most general solution. I definitely felt that I could have solved this problem if I knew how to get you to call some of my code before calling _validate_mapping. (in fact monkey-patching _validate_mapping is how I "fixed" the problem)

**Comment by unbrice** [comment]
I was trying to integrate strictyaml with Pydantic and thought this would be useful.

For my use case, what would work is to use the `GetterDict` (which may not have existed when this thread was started) and `from_obj`. I was thinking of adding a method to `GetterDict`:
```python
def field_location(self, field) -> Union[str, 'Loc']:
  return field
```
that subclass can override to mix in their location informations.

If this looks good to you I can propose a PR. I suspect it would also work for OP.

**Comment by samuelcolvin** [comment]
this is still something, I'd love to have. I'll continue to think about the best way to achieve it.

**Comment by esikora** [comment]
I have applied the approach outlined above in order to track file, line and column information in a Pydantic model using Pydantic 2.9 / 2.10. Since I stumbled over a few challenges, I would like to describe how these were resolved. Improvements are welcome.

The following solution considers model classes defined with `model_config = ConfigDict(extra='forbid')`.

Initially, location information is obtained using a customized yaml loader:
```python
class MyLoader(SafeLoader):
   def __init__(self, stream):
       super().__init__(stream)

   def construct_mapping(self, node: MappingNode, deep: bool=False):
       mapping = super().construct_mapping(node, deep=deep)
       mapping['_loc_file'] = node.start_mark.name
       mapping['_loc_line'] = node.start_mark.line + 1
       mapping['_loc_col'] = node.start_mark.column + 1
       return mapping
```

Model elements that shall have location information subclass the following class. Therein, location information is stored using PrivateAttr. A wrap validator is used to extract and remove the location from that raw data and to set the private attributes.
```python
class MyBaseModel(BaseModel):
    _loc_file: Optional[str] = PrivateAttr(default=None)
    _loc_line: Optional[int] = PrivateAttr(default=None)
    _loc_col: Optional[int] = PrivateAttr(default=None)

    @model_validator(mode="wrap")
    def set_location(cls, values, handler):
        if isinstance(values, dict):
            loc_file = values.pop('_loc_file', None)
            loc_line = values.pop('_loc_line', None)
            loc_col = values.pop('_loc_col', None)
        model = handler(values)
        model._loc_file = loc_file
        model._loc_line = loc_line
        model._loc_col = loc_col
        return model
```
This approach resolves issue related to a custom `__init__` function. The custom `__init__` had caused validations that require a `ValidationInfo.context` argument not to work any more. As far as I could figure out, Pydantic did not pass the context objects to these validations (or maybe did not call them at all).

As mentioned in https://github.com/pydantic/pydantic/issues/1273#issuecomment-593500316, the custom loader inserts location information into any Dict-like model elements. As this caused validation errors, I strip the information using a BeforeValidator:
```python
def strip_location(v: Any):
    if isinstance(v, dict):
        v.pop('_loc_file', None)
        v.pop('_loc_line', None)
        v.pop('_loc_col', None)
        return v

class MyClass(MyBaseModel):
    my_dict_attribute: Optional[
        Annotated[
            Dict[str, MyElementClass],
            BeforeValidator(strip_location)
        ]] = None
```
